Don’t ack if unable to remove failing replica

dnhatn opened the pull request:
Today when a replicated write operation fails to execute on a replica, the primary will reach out to the master to fail that replica (and mark it stale). We then won't ack that request until the master removes the failing replica; otherwise, we will lose the acked operation if the failed replica is still in the in-sync set. However, if a node with the primary is shutting down, we might ack such request even though we are unable to send a shard-failure request to the master. This happens because we ignore NodeClosedException which is triggered when the ClusterService is being closed.

Closes #39467

/cc @bleskes @martijnvg @jasontedor

dnhatn committed:
Don’t ack if unable to remove failing replica

elasticmachine commented:
Pinging @elastic/es-distributed

dnhatn commented on server/src/test/java/org/elasticsearch/discovery/ClusterDisruptionIT.java:
```diff
@@ -436,4 +442,54 @@ public void testIndicesDeleted() throws Exception {
         assertFalse(client().admin().indices().prepareExists(idxName).get().isExists());
     }
 
+    public void testRestartPrimaryNodeWhileIndexing() throws Exception {
```
I will combine this newly added test to `testAckedIndexing` in a follow-up.

ywelsch replied:
was this test able to expose the bug without the fix?

ywelsch commented on server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -204,13 +202,18 @@ public String toString() {
+            primary.failShard(message, failure);
+        } else {
+            // these can occur if the node is shutting down and are okay any other exception here is not expected and merits investigation.
+            assert failure instanceof NodeClosedException || failure instanceof TransportException : failure;
```
@bleskes mentioned that the TransportException here should be coming from https://github.com/elastic/elasticsearch/blob/022726011ce1d9e4748fd39293a3aeae6954b3ac/server/src/main/java/org/elasticsearch/transport/TransportService.java#L645-L649

We should verify this here with an assertion and (I think in a follow-up) look into throwing a more appropriate exception in TransportService. Possible options are `AlreadyClosedException` or a custom subclass of `TransportException`.

dnhatn replied:
I pushed https://github.com/elastic/elasticsearch/pull/39584/commits/6afb094f04d21f2f30c601a4beea9fa25de06c6c.

ywelsch commented on server/src/test/java/org/elasticsearch/discovery/ClusterDisruptionIT.java:
```diff
@@ -436,4 +442,54 @@ public void testIndicesDeleted() throws Exception {
+        for (ShardRouting shardRouting : clusterState.routingTable().allShards(index)) {
+            if (shardRouting.primary()) {
+                String nodeName = clusterState.nodes().get(shardRouting.currentNodeId()).getName();
+                internalCluster().restartNode(nodeName, new InternalTestCluster.RestartCallback());
```
perhaps just restart a random node instead of explicitly the one with the primary?

dnhatn replied:
Yep, I pushed https://github.com/elastic/elasticsearch/pull/39584/commits/ab65f4b14ae8af85708e8b4340fd72599546f803

dnhatn committed:
restart random data node

dnhatn committed:
assert transport exception

dnhatn committed:
Merge branch 'master' into primary-shutdown

dnhatn commented:
@ywelsch Thanks for looking. It's ready again.

ywelsch commented on server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -204,13 +203,23 @@ public String toString() {
+            (failure instanceof TransportException && "TransportService is closed stopped can't send request".equals(failure.getMessage()));
+        final String message;
+        if (nodeIsClosing) {
+            message = String.format(Locale.ROOT, "primary node [%s] is shutting down while failing replica shard", primary.routingEntry());
```
perhaps change to "node with primary [%s] is shutting ..."

dnhatn replied:
Pushed https://github.com/elastic/elasticsearch/pull/39584/commits/6b56fbcee0171e995a01ac7f1ba288963be44d41.

ywelsch commented on server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -204,13 +203,23 @@ public String toString() {
+            (failure instanceof TransportException && "TransportService is closed stopped can't send request".equals(failure.getMessage()));
+        final String message;
+        if (nodeIsClosing) {
+            message = String.format(Locale.ROOT, "primary node [%s] is shutting down while failing replica shard", primary.routingEntry());
```
add a comment here as well that it's not necessary to fail shard as the node is shutting down anyway

dnhatn replied:
Added https://github.com/elastic/elasticsearch/pull/39584/commits/6b56fbcee0171e995a01ac7f1ba288963be44d41

ywelsch reviewed (approved):
LGTM.

dnhatn committed:
Merge branch 'master' into primary-shutdown

dnhatn committed:
wording + comment

dnhatn commented:
@ywelsch Thanks!

dnhatn commented:
@jpountz I have backported this to 6.6.2 in https://github.com/elastic/elasticsearch/commit/799491fb8c9dd41bdcdfbe5333c213eaeebb180a.

dnhatn commented:
@jpountz @ywelsch I have reverted this change in 6.6.2 for we need to adjust the way we check if the node with primary is shutting down. It's fine if this does not make 6.6.2 (I've removed the label).

jpountz commented:
Thanks for letting me know!
